clients/cobol-lsp-vscode-extension/src/__tests__/services/copybook/download/CopybookDownloaderForE4E.spec.ts

```diff
@@ -0,0 +1,159 @@
+ * The term "Broadcom" refers to Broadcom Inc. and/or its subsidiaries.
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Broadcom, Inc. - initial API and implementation
+ */
+import { CopybookDownloaderForE4E } from "../../../../services/copybook/downloader/CopybookDownloaderForE4E";
+import * as path from "path";
+import * as fs from "node:fs";
+import { e4eErrorResponseMock, e4eResponseMock } from "./getE4EMock.utility";
+import e4eTypes = require("../../../../type/e4eApi.d");
+
+jest.mock("node:fs", () => ({
+  promises: {
+    writeFile: jest.fn(),
+  },
+  mkdirSync: jest.fn(),
+  existsSync: jest.fn(),
+}));
+
+import * as vscode from "vscode";
+const e4eDownloader = new CopybookDownloaderForE4E("/storagePath", undefined);
+
+describe("e4e copybook downloader tests", () => {
+  beforeEach(() => {
+    jest.clearAllMocks();
+  });
+  it("checks copybook downloaded into correct path", () => {
+    expect(
+      (CopybookDownloaderForE4E as any).getCopybookPath(
+        "Instance.Instance",
+        "pgm",
+        "/storagePath",
+        "Copy.cpy",
+      ),
+    ).toEqual(
+      path.join(
+        "/storagePath",
+        ".e4e",
+        ".copybooks",
+        "Instance.Instance",
+        "pgm",
+        "Copy",
+      ),
+    );
+    expect(fs.mkdirSync).toHaveBeenCalled();
+  });
+  it("checks not to try to download any if member or element not available in e4e", () => {
+    e4eDownloader.downloadCopybookE4E(
+      { name: "NoCopybook", dialect: "COBOL" },
+      "document-uri",
+      e4eResponseMock,
+    );
+
+    const spyDownloadDataset = jest.spyOn(e4eDownloader, "downloadDatasetE4E");
+    const spyDownloadElement = jest.spyOn(e4eDownloader, "downloadElementE4E");
+    expect(spyDownloadDataset).not.toHaveBeenCalled();
+    expect(spyDownloadElement).not.toHaveBeenCalled();
+  });
+  it("check download performed only for member when copyboook matches member and element", () => {
+    const spyDownloadDataset = jest.spyOn(e4eDownloader, "downloadDatasetE4E");
+    const spyDownloadElement = jest.spyOn(e4eDownloader, "downloadElementE4E");
+    e4eDownloader.downloadCopybookE4E(
+      { name: "copybook", dialect: "COBOL" },
+      "document-uri",
+      e4eResponseMock,
+    );
+    expect(spyDownloadDataset).toHaveBeenCalledWith(e4eResponseMock, {
+      dataset: "dataset",
+      member: "copybook",
+    });
+    expect(spyDownloadElement).not.toHaveBeenCalled();
+  });
+  it("check download performed only for element when no member matches", () => {
+    const spyDownloadDataset = jest.spyOn(e4eDownloader, "downloadDatasetE4E");
+    const spyDownloadElement = jest.spyOn(e4eDownloader, "downloadElementE4E");
+    const responseMock = Object.assign({}, e4eResponseMock);
+    responseMock.members = [];
```

Comment:
why? just call the endevor element in e4eResponseMock something else... or add an extra one to have both conflicting and non-conflicting names.